lhotse/dataset/sampling/dynamic_bucketing.py

```diff
@@ -427,38 +448,51 @@ def is_ready(bucket: Deque[Cut]):
+                sampling_bucket = ready_buckets[bucket_idx]
+
                 # Apply random shuffling if requested: we'll shuffle the items present within the bucket.
                 maybe_shuffled = sampling_bucket
                 indexes_used = []
                 if self.shuffle:
                     maybe_shuffled = pick_at_random(
-                        maybe_shuffled, rng=self.rng, out_indexes_used=indexes_used
+                        maybe_shuffled,
+                        rng=self.bucket_rng,
+                        out_indexes_used=indexes_used,
                     )
-                # Sample one batch from that bucket and yield it to the caller.
+                else:
+                    maybe_shuffled = pick_at_sequential(
+                        maybe_shuffled,
+                        out_indexes_used=indexes_used,
+                    )
+
+                # Sample world_size batches from that bucket and yield it to the caller.
+                # Force More similar mean batch duration across nodes with DynamicBucketingSampler in multi-GPU training
                 batcher = DurationBatcher(
                     maybe_shuffled,
                     constraint=self.constraint.copy(),
                     diagnostics=self.diagnostics,
                 )
-                batch = next(iter(batcher))
-                if isinstance(batch, tuple):
-                    batch_size = len(batch[0])
-                else:
-                    batch_size = len(batch)
-                yield batch
+                batcher = iter(batcher)
+
+                total_batch_size = 0
+                for _ in range(self.world_size):
+                    batch = next(batcher)
+                    if isinstance(batch, tuple):
+                        batch_size = len(batch[0])
+                    else:
+                        batch_size = len(batch)
+                    total_batch_size += batch_size
+                    yield batch
+
                 # Remove sampled cuts from the bucket.
-                if indexes_used:
-                    # Shuffling, sort indexes of yielded elements largest -> smallest and remove them
-                    indexes_used.sort(reverse=True)
-                    for idx in indexes_used:
-                        del sampling_bucket[idx]
-                else:
-                    # No shuffling, remove first N
-                    for _ in range(batch_size):
-                        sampling_bucket.popleft()
+                # Shuffling, sort indexes of yielded elements largest -> smallest and remove them
+                indexes_used.sort(reverse=True)
+                for idx in indexes_used:
+                    del sampling_bucket[idx]
+
                 # Fetch new cuts and add them to appropriate buckets.
-                self._collect_cuts_in_buckets(batch_size)
+                self._collect_cuts_in_buckets(total_batch_size * self.world_size)
```

Comment:
we need collect more cuts here, It is often encountered that there is insufficient data in buckets.  @pzelasko Do you have any suggestion?

Reply:
sample batches from the some bucket(word_size times), then get batches[rank].